### PR TITLE
Added identity header support

### PR DESCRIPTION
**Pull request by jroper**
Passes the identity as described in https://github.com/linkerd/linkerd2/issues/3756.

So firstly, until yesterday, I had never seen a single line of Rust code before. I know Rust uses explicit memory management, and I'm sure I've got some things wrong because really, I have no idea what I'm doing.

I've probably inserted my identity header logic in the wrong place in the stack, and I'm not sure if the identity parameter gets wired up correctly at the right time or not.

I haven't implemented an integration test because I'm not sure where I should put it, or whether there is a good test to model my tests from. But the tests that are needed here are:

* Validate that the identity header is passed when present.
* Validate that any client supplied identity headers, both for TLS and non TLS clients, are stripped.

### CODE REVIEW

**Comment by olix0r** [comment]
@jroper Thanks! At a quick glance, this change looks good, but I'll give it a deeper look in a bit. In the meantime, you'll notice that the PR is failing two checks:

1. You can run `cargo fmt` to run rustfmt on this, which will resolve the formatting check.
2. This change probably warrants [DCO sign-off](https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#developer-certificate-of-origin). You can either amend your commits or simply make a statement in this PR to the effect of "I agree to the DCO for all the commits in this PR."

**Comment by jroper** [comment]
@olix0r I'll update. I would like to write tests for it though before you merge, is there any particular test that would be good to base my tests on?

**Comment by jroper** [comment]
I'm part way through writing a test, but I haven't worked out how to, in the integration tests, get the proxy to not pass through the TLS connection, here's my code:

```rust
    let client_id = "foo.ns1.serviceaccount.identity.linkerd.cluster.local";
    let client_id_svc = identity::Identity::new("foo-ns1", client_id.to_string());
    let app_id = "bar.ns1.serviceaccount.identity.linkerd.cluster.local";
    let app_id_svc = identity::Identity::new("bar-ns1", app_id.to_string());

    let srv = server::http1()
        .route_fn("/check-identity", |req| -> Response<Bytes> {
            [snip]
        })
        .run()
        .await;

    let srv_addr = srv.addr;
    let dst = format!("identity.test.svc.cluster.local:{}", srv_addr.port());
    let ctrl = controller::new();
    let _profile_in = ctrl.profile_tx_default(&dst, "identity.test.svc.cluster.local");
    let ctrl = ctrl
        .run()
        .instrument(tracing::info_span!("ctrl", "inbound"))
        .await;
    let proxy = proxy::new()
        .controller(ctrl)
        .inbound(srv)
        .identity(app_id_svc.service().run().await)
        .run_with_test_env(app_id_svc.env)
        .await;

    tracing::info!("tls request to {}", proxy.inbound);
    let tls_client = client::http1_tls(
        proxy.inbound,
        "identity.test.svc.cluster.local",
        client::TlsConfig::new(client_id_svc.client_config, client_id),
    );
```

This is what I see:

```
[     0.018291s] DEBUG send_req{req=Request { method: GET, uri: https://identity.test.svc.cluster.local/ready, version: HTTP/1.1, headers: {}, body: b"" }}: linkerd_app_integration::client: request headers={}
[     0.019288s] DEBUG proxy{test=tests::identity::identity_header_passing:proxy}: linkerd_tls::client: Peer does not support TLS reason=loopback
[     0.019760s] DEBUG proxy{test=tests::identity::identity_header_passing:proxy}:inbound:accept{client.addr=127.0.0.1:39368}:server{port=48897}: linkerd_app_inbound::prevent_loop: addr=127.0.0.1:48897 self.port=41683
[     0.021630s] DEBUG proxy{test=tests::identity::identity_header_passing:proxy}:inbound:accept{client.addr=127.0.0.1:39368}:server{port=48897}:proxy: linkerd_tls::server: Peeked bytes from TCP stream sz=280
[     0.021747s] DEBUG proxy{test=tests::identity::identity_header_passing:proxy}:inbound:accept{client.addr=127.0.0.1:39368}:server{port=48897}:proxy: linkerd_tls::server: Identified foreign SNI sni=foo.ns1.serviceaccount.identity.linkerd.cluster.local
[     0.021852s] DEBUG proxy{test=tests::identity::identity_header_passing:proxy}:inbound:accept{client.addr=127.0.0.1:39368}:server{port=48897}:proxy: linkerd_app_inbound::require_identity: port=48897 tls=Some(Passthru { sni: ServerId("foo.ns1.serviceaccount.identity.linkerd.cluster.local") }) id_required=false
[     0.022081s] DEBUG proxy{test=tests::identity::identity_header_passing:proxy}:inbound:accept{client.addr=127.0.0.1:39368}:server{port=48897}:proxy: linkerd_detect: DetectResult protocol=None elapsed=58.076µs
[     0.023056s] DEBUG test_server{version=Http1 addr=127.0.0.1:48897 test=tests::identity::identity_header_passing}:conn{addr=127.0.0.1:40108}: linkerd_app_integration::server: accepting connection tls=false
[     0.023521s] ERROR test_server{version=Http1 addr=127.0.0.1:48897 test=tests::identity::identity_header_passing}:conn{addr=127.0.0.1:40108}: linkerd_app_integration::server: support/server error: invalid HTTP method parsed
```

I don't know why it's using Passthru, not sure how to configure it not to.

**Comment by olix0r** [comment]
@jroper we haven't forgotten this, but we've been focusing on getting some high-priority issues on main addressed. it's definitely awkward to write tests in this repo that depend on TLS infrastructure at the moment, but we can probably help fix that. If you're getting stuck, it might be preferable to try to address this with integration tests in the linkerd2 repo (to augment unit tests in this repo).

cc @hawkw

**Comment by jroper** [comment]
I've added one test, it tests that the l5d-identity header is stripped if a client tries to spoof it. I couldn't work out how to get TLS working, I did work out why it was passing through before (I was using the wrong name for SNI), but then it failed because it was unable to verify the certificate chain.

I have tested this manually though, this is what a request to [this echo http service](https://hub.docker.com/r/mendhak/http-https-echo) returns:

```
{
  "path": "/",
  "headers": {
    "host": "echo",
    "user-agent": "curl/7.76.1-DEV",
    "accept": "*/*",
    "l5d-dst-canonical": "echo.default.svc.cluster.local:80",
    "l5d-identity": "other.default.serviceaccount.identity.linkerd.cluster.local"
  },
  "method": "GET",
  "body": "",
  "fresh": false,
  "hostname": "echo",
  "ip": "::ffff:127.0.0.1",
  "ips": [],
  "protocol": "http",
  "query": {},
  "subdomains": [],
  "xhr": false,
  "os": {
    "hostname": "echo-86cc7bcf7c-gqvkr"
  },
  "connection": {}
}
```

**Comment by jroper** [comment]
Is there anything else I can do to help this along?

**Comment by olix0r** [comment]
@jroper Excuse the delay, I've been a bit distracted with non-proxy work for the past week+. I plan to spend some time catching up on the proxy backlog today and next week.

You might also be interested in the policy design we just published, which covers setting headers like those you've added in this PR https://github.com/linkerd/polixy/blob/main/DESIGN.md#httpgrpc-headers-

**Comment by jroper** [comment]
So, that doc mentions a `l5d-client-id` header, which is what I've just implemented, though I called it `l5d-identity`. Was that documented there as an intended future feature?

**Comment by olix0r** [comment]
@jroper The linked policy work is planned for the upcoming stable release. I think it makes sense to include as part of the policy work, especially, because we're adding new types of CRD resources to support policy enforcement, and this feature helps to surface that to applications.

This branch looks great. I really appreciate you taking this on. I'm going to start building on this branch to address some of the small merge conflicts we've just made and to bring this more closely in line with where we see this going in the upcoming stable release.

**Comment by olix0r** [comment]
I made a few small changes:

* Consolidated everything into `linkerd::app::inbound::http::set_identity_header` -- since this module isn't reused there's not really any advantage to having it in common code. Similarly, there's no need to define constants in common code.
* With this, the change doesn't have to be generic/configurable over its header name.
* Changed the header name to `l5d-client-id`.
* Some other optimizations including:
  * Avoid duplicate header lookups -- `HeaderMap::insert` clears prior values
  * We create a `HeaderValue` once when the service is created rather than on each request.

**Review by olix0r — APPROVED** [review]

**Comment by olix0r** [comment]
Thanks again for your help @jroper! I'll try to get this included in this week's edge release (tomorrow). Worst case, it will be included in next week's edge release.

**Comment by jroper** [comment]
Thanks so much! It's been a pleasure to contribute! And now I have a basic knowledge of Rust!

File: linkerd/app/integration/src/tests/identity.rs [hunk 1]
```diff
@@ -468,3 +468,51 @@ mod require_id_header {
...
+            == http::StatusCode::OK
+    );
+
+    assert_eq!(
+        client
+            .request(
+                client
+                    .request_builder("/check-identity")
+                    .method("GET")
+                    .header("l5d-identity", "spoof")
```
Comment (dragonsinth) [inline linkerd/app/integration/src/tests/identity.rs:511]: What this supposed to be `l5d-client-id` also?

**Pull request merged**